### DISCUSSION TITLE
Overlapping Overload with Literals Question

### DISCUSSION THREAD

**Discussion by to-ph** [microsoft/pyright#2610] [Q&A]
Hello! I've got a quick question about overloading in a scenario like this:

```python

@overload
def greet(s: Literal["hello"]) -> Literal["hi!"]:
    ...

@overload
def greet(s: str) -> Literal["what?"]:
    ...

def greet(s: str):
    if s == "hello":
        return "hi!"
    else:
        return "what?"
```

The `str` in the second `greet` overlaps the literal `"hello"`, and pyright (correctly) complains about an incompatible return type. Is there any way to make this scenario valid, or is it just a limit with python's type system that I would need to work around?

Thanks!

**Comment by erictraut** [comment]
I guess you could consider it a limitation of the type system. There's no way to say "any `str` that is not `Literal["hello"]`, so it's not possible to make the two overloaded signatures distinct (not overlapping).

Here are some options:
1. Change the return type of the second overload (and the implementation) to `str`.
2. Suppress the `reportOverlappingOverloads` diagnostic check for this file (or use `# type: ignore` to suppress the diagnostic on a single line).
